Re-implement Nix CI workflow on GitHub instead of CircleCI

btlogy opened the pull request.

hacklschorsch committed:
Update PyOpenSSL to the second-latest version

The latest version 404s ?

btlogy committed:
Implement Nix CI workflow on GitHub instead of CircleCI

Signed-off-by: Benoit Donneaux <benoit@leastauthority.com>

btlogy committed:
Update the Nix flake to follow only nixos-24_05

Signed-off-by: Benoit Donneaux <benoit@leastauthority.com>

btlogy committed:
Our flake can only support Python 3.10 and 3.11

- ipython-8.24.0 not supported for interpreter python3.9
- magic-wormhole-transit-relay-0.2.1 not supported for interpreter python3.12

Signed-off-by: Benoit Donneaux <benoit@leastauthority.com>

btlogy committed:
Update flake to use nixos-24_11

Signed-off-by: Benoit Donneaux <benoit@leastauthority.com>

btlogy committed:
FIXME: skip flake check for now becasue of imcompatibility

- tkinter-7.3.12 not supported for interpreter pypy3.9

Signed-off-by: Benoit Donneaux <benoit@leastauthority.com>

hacklschorsch commented:
I have some updates for tahoe-lafs dependencies in here: https://github.com/PrivateStorageio/ZKAPAuthorizer/pull/481

hacklschorsch committed:
Drop our vendored pycddl, upstream has our fixes now

hacklschorsch committed:
Update nixpkgs in github workflow definition too

hacklschorsch committed:
Klein asks for hypothesis to run its tests

hacklschorsch committed:
Use the nixpkgs-provided klein

hacklschorsch committed:
Use the nixpkgs-provided pyopenssl

hacklschorsch committed:
Use the nixpkgs-provided twisted

hacklschorsch committed:
Use nixpkgs-provided txtorcon

hacklschorsch committed:
Use nixpkgs-provided txi2p-tahoe

hacklschorsch committed:
Forgot to remove service-identity.nix which belonged to our pyopenssl fix

hacklschorsch committed:
Spell txi2p correctly

hacklschorsch committed:
Also run CI for Python 3.12

hacklschorsch committed:
Allow Python 3.12

hacklschorsch committed:
Patch _version.py

Courtesy of the ZKAPAuthorizer nix expressions

hacklschorsch committed:
The missing version fails a unit test - does this help?

hacklschorsch commented:
Much better than packaging things IRL ;)

hacklschorsch committed:
Don't test Python 3.10 on NixOS 24.11

... it's old (3.12 is the default), it's not in
the NixOS binary caches, some dependency fails building
and fixing it feels like picking over bones

hacklschorsch committed:
Remove check step

... it has served its purpose (development) and is using runner time

hacklschorsch commented:
I saw that, contrary to what I had written in my commit message 68f706e689b1bea013c9c08af7eb71b60fe2c872, also Python 3.10 succeeds, for which a bunch of dependencies have to be built because they are not in the NixOS cache anymore.

But really IDK what we learn from making old Python versions work on NixOS - I think only testing the current stable channel's default (here: 3.12) is fine.

hacklschorsch committed:
Don't test obsolete Python 3.11, don't depend on check phase

hacklschorsch committed:
Does this create the environment variables?

Why is this so complicated?  I thought the nixpkgs were
defined in flake.lock?

hacklschorsch committed:
Setting the env vars already requires Nix

hacklschorsch committed:
Remove input overriding

hacklschorsch committed:
Try out what the default number of cores does on GitHub Actions

hacklschorsch committed:
Restore old / more generic version guard

... so we have less places where we have to change versions

hacklschorsch committed:
Have fewer outdated comments

hacklschorsch committed:
Revert "Try out what the default number of cores does on GitHub Actions"

This reverts commit 0c0b83536685c1c21d7591d17bed364c732abf95.

Not setting cores by hand made the build take 5m31s instead of 2m46s.
Undoing that change.

hacklschorsch committed:
Remove NixOS jobs from CircleCI

... they run much better in GitHub Actions

btlogy commented:
> @ben do you see a reason why we shouldn't remove Python 3.11 as well?

My understanding of the implications is too weak I'm afraid.
I would generally think it's good to stick to one stable/well supported version to checks PRs (e.g.: Python 3.12 I guess), and try to foreseen possible upcoming issues in the next version (e.g.: Python 3.13?)

Though, his is sometime problematic for LTS distros which keeps using stable version longer than what upstream keeps supporting...

But I've not yet a clear idea what this means for NixOS honestly.

meejah commented:
> Though, his is sometime problematic for LTS distros which keeps using stable version longer than what upstream keeps supporting...

For reference:
- Python 3.12 is supported until 2028-10. 
- Python 3.11 is supported until 2027-10.

btlogy commented:
> > Though, this is sometime problematic for LTS distros which keeps using stable version longer than what upstream keeps supporting...
> 
> For reference:
> 
>     * Python 3.12 is supported until 2028-10.
> 
>     * Python 3.11 is supported until 2027-10.

So we have time.

But what I meant is that some distro support things like Python longer than the upstream developers.
For instance, I was looking at the EOL for this OracleLinux 8 to see how long they would support Python 3.9...
And look what I've found [there](https://docs.oracle.com/en/operating-systems/oracle-linux/product-lifecycle/OL-PRODUCT-LIFECYCLE.pdf):

> python 3.6 EOL July 2029

:scream:
